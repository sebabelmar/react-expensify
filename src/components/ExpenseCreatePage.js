import React from 'react';

const AddExpensesPage = () => (
  <div>
    This is the add expenses page.
  </div>
);

export default AddExpensesPage;
