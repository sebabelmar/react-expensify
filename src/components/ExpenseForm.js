import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calendarFocused: false,
    error: ''
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }
  onAmountChange = (e) => {
    e.persist();
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount: e.target.value }));
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.amount || !this.state.description) {
      this.setState(() => ({
        error: 'Amount and description are required...'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      const expense = {
        seba: 'yeah'
      };
      this.props.onSubmit(expense);
    }
  }
  render() {
    return (
      <div>
        { this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Subimt</button>
        </form>
      </div>
    );
  }
}
