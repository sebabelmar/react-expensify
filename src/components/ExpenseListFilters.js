import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

// Refactor to use mapDispatchToProps so we can create spies and pass them as mapStateToProps
// Dispachers should be call with the expected parameters
// state should change filters {...}

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }
  onTextChange = (e)=>{
    this.props.setTextFilter(e.target.value);
  }
  onSortChange = (e) => {
    e.target.value === 'amount'
    ? this.props.sortByAmount()
    : this.props.sortByDate();
  }
  render() {
    return (
      <div>
      <input
        type="text"
        value={this.props.filters.text}
        onChange={this.onTextChange}
      />
      <select
        value={this.props.filters.sortBy}
        onChange={this.onSortChange}>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker
        startDate={this.props.filters.startDate}
        endDate={this.props.filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        startDateId="id"
        endDateId="id"
        showClearDates
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
      </div>
    );
  }
}

const mapDispatchToPropsTo = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

const mapStateToProps = (state) => ({
  filters: state.filters
});

export default connect(mapStateToProps, mapDispatchToPropsTo)(ExpenseListFilters);
