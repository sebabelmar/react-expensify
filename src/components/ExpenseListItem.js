import React from 'react';
import { removeExpense } from '../actions/expenses';
import { connect } from 'react-redux';

const ExpenseListItem = (props) => {
  const {id, description, amount, createdAt} = props;
  return (
    <div>
      <h3>{description}</h3>
      <p>${amount} - {createdAt}</p>
      <button onClick={()=>{
        props.dispatch(removeExpense(id));
      }}>Remove</button>
    </div>
  );
};

export default connect()(ExpenseListItem);
