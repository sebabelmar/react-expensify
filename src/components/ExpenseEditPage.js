import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';

const EditExpensePage = () => {
  return (
    <div>
      <ExpenseForm />
    </div>
  );
};

export default connect()(EditExpensePage);
