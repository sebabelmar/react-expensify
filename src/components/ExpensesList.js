import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpensesList = (props) => (
  <div>
    <h1>Expenses List</h1>
    {
      props.expenses.map((expense) =>
        (<ExpenseListItem {...expense } key={expense.id}/>))
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpensesList);
