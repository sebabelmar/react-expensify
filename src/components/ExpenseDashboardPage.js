import React from 'react';
import ExpensesList from './ExpensesList';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesList />
  </div>
);

export default ExpenseDashboardPage;
