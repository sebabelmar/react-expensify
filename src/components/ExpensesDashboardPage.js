import React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpensesDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpensesSummary />
    <ExpensesList />
  </div>
);

export default ExpensesDashboardPage;
