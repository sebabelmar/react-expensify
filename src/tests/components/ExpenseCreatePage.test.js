import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensesPage } from '../../components/ExpenseCreatePage';
import expenses from '../fixtures/expenses';

let addExpense;
let history;
let wrapper;

beforeEach(() => {
  addExpense = jest.fn();
  history = {push: jest.fn()};
  wrapper = shallow(<AddExpensesPage addExpense={addExpense} history={history}/>);
});

test('should render AddExpensesPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle on submit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);

  expect(addExpense).toHaveBeenLastCalledWith(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(wrapper).toMatchSnapshot();
});
