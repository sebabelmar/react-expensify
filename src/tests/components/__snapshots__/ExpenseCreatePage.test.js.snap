// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle on submit 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render AddExpensesPage correctly 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
